cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageCollapser.java

```diff
@@ -0,0 +1,178 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.metadata;
+
+import co.cask.cdap.data2.metadata.lineage.AccessType;
+import co.cask.cdap.data2.metadata.lineage.Relation;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.metadata.lineage.CollapseType;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+import org.apache.twill.api.RunId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Collapses {@link Relation}s based on {@link CollapseType}
+ */
+public class LineageCollapser {
+  private static final Logger LOG = LoggerFactory.getLogger(LineageCollapser.class);
+
+  /**
+   * Collapse {@link Relation}s based on {@link CollapseType}
+   * @param relations lineage relations
+   * @param collapseTypes fields to collapse relations on
+   * @return collapsed relations
+   */
+  public static Set<CollapsedRelation> collapseRelations(Iterable<Relation> relations,
+                                                         Set<CollapseType> collapseTypes) {
+    Set<CollapsedRelation> collapsedRelations = new HashSet<>();
+
+    Multimap<CollapseKey, Relation> multimap = HashMultimap.create();
+    for (Relation relation : relations) {
+      multimap.put(getCollapseKey(relation, collapseTypes), relation);
+    }
+    LOG.trace("Collapsed relations: {}", multimap.asMap());
+
+    for (Map.Entry<CollapseKey, Collection<Relation>> collapsedEntry : multimap.asMap().entrySet()) {
+      Id.NamespacedId data = collapsedEntry.getKey().data;
+      Id.Program program = collapsedEntry.getKey().program;
+
+      Set<AccessType> accessTypes = new HashSet<>();
+      Set<RunId> runs = new HashSet<>();
+      Set<Id.NamespacedId> components = new HashSet<>();
+
+      for (Relation relation : collapsedEntry.getValue()) {
+        accessTypes.add(relation.getAccess());
+        runs.add(relation.getRun());
+        components.addAll(relation.getComponents());
+      }
+      collapsedRelations.add(toCollapsedRelation(data, program, accessTypes, runs, components));
+    }
+    return collapsedRelations;
+  }
+
+  private static CollapseKey getCollapseKey(Relation relation, Set<CollapseType> collapseTypes) {
+    CollapseKeyBuilder builder = new CollapseKeyBuilder(relation.getData(), relation.getProgram());
+    if (!collapseTypes.contains(CollapseType.ACCESS)) {
+      builder.setAccess(relation.getAccess());
+    }
+    if (!collapseTypes.contains(CollapseType.RUN)) {
+      builder.setRun(relation.getRun());
+    }
+    if (!collapseTypes.contains(CollapseType.COMPONENT)) {
+      builder.setComponents(relation.getComponents());
+    }
+    return builder.build();
+  }
+
+  private static final class CollapseKeyBuilder {
```

Comment:
Since `CollapseKey` is already a nested private class, I didn't want to double nest.